analysis_termination.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+)
+
+// JobKiller is responsible for killing jobs either in HTCondor or in K8s.
+type JobKiller struct {
+	K8sEnabled bool        // whether or not the VICE apps are running k8s
+	AppsBase   string      // base URL for the apps service
+	appExposer *AppExposer // base URL for the app-exposer serivce
+	dedb       *sql.DB
+	jslURL     *url.URL
+}
+
+func NewJobKiller(k8sEnabled bool, appsBase, aeBase string, dedb *sql.DB, jslURL *url.URL) (*JobKiller, error) {
+	appExposer, err := NewAppExposer(aeBase)
+	if err != nil {
+		return nil, err
+	}
+	return &JobKiller{
+		K8sEnabled: k8sEnabled,
+		AppsBase:   appsBase,
+		appExposer: appExposer,
+		dedb:       dedb,
+		jslURL:     jslURL,
+	}, nil
+}
+
+// KillJob uses either the apps or app-exposer APIs to kill a VICE job.
+func (j *JobKiller) KillJob(ctx context.Context, job *Job) error {
+	if j.K8sEnabled {
+		return j.killK8sJob(ctx, job)
+	}
+	return j.killCondorJob(ctx, job.ID, job.User)
+
+}
+
+// killCondorJob uses the provided API at the base URL to kill a running job. This
+// will probably be to the apps service. jobID should be the UUID for the Job,
+// typically returned in the ID field by the analyses service. The username
+// should be the short username for the user that launched the job.
+func (j *JobKiller) killCondorJob(ctx context.Context, jobID, username string) error {
+	return j.appExposer.StopAnalyses(ctx, jobID, username)
+}
+
+func (j *JobKiller) requestAnalysisTermination(ctx context.Context, job *Job) error {
+	return j.appExposer.VICESaveAndExit(ctx, job)
+}
```

Comment:
I like the idea of having this stuff encapsulated in the `app-exposer` client. 👍